nestkernel/sonata_connector.cpp

```diff
@@ -0,0 +1,666 @@
+  }
+  return group;
+}
+
+H5::Group*
+SonataConnector::open_group_( const H5::Group* group, const std::string& grp_name )
+{
+  H5::Group* subgroup = nullptr;
+  try
+  {
+    subgroup = new H5::Group( group->openGroup( grp_name ) );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open HDF5 group " + grp_name + " in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+  return subgroup;
+}
+
+
+void
+SonataConnector::open_required_dsets_( const H5::Group* pop_grp )
+{
+  try
+  {
+    src_node_id_dset_ = pop_grp->openDataSet( "source_node_id" );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open source_node_id dataset in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+
+  try
+  {
+    tgt_node_id_dset_ = pop_grp->openDataSet( "target_node_id" );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open target_node_id dataset in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+
+  try
+  {
+    edge_type_id_dset_ = pop_grp->openDataSet( "edge_type_id" );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open edge_type_id dataset in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+
+  // Consistency checks
+  const auto num_tgt_node_ids = get_nrows_( tgt_node_id_dset_ );
+
+  // Ensure that target and source population have the same size
+  if ( num_tgt_node_ids != get_nrows_( src_node_id_dset_ ) )
+  {
+    throw KernelException(
+      "target_node_id and source_node_id datasets in " + cur_fname_ + " must be of the same size" );
+  }
+
+  // Ensure that edge_type_id dataset size is consistent with the number of target node ids
+  if ( num_tgt_node_ids != get_nrows_( edge_type_id_dset_ ) )
+  {
+    throw KernelException( "target_node_id and edge_type_id datasets in " + cur_fname_ + " must be of the same size" );
+  }
+}
+
+void
+SonataConnector::try_open_edge_group_dsets_( const H5::Group* edge_grp )
+{
+  // TODO: Currently only works if the edge file has a single edge group
+
+  weight_dataset_exist_ = H5Lexists( edge_grp->getId(), "syn_weight", H5P_DEFAULT ) > 0;
+  delay_dataset_exist_ = H5Lexists( edge_grp->getId(), "delay", H5P_DEFAULT ) > 0;
+
+  if ( weight_dataset_exist_ )
+  {
+    try
+    {
+      syn_weight_dset_ = edge_grp->openDataSet( "syn_weight" );
+    }
+    catch ( const H5::Exception& e )
+    {
+      throw KernelException( "Could not open syn_weight dataset in " + cur_fname_ + ": " + e.getDetailMsg() );
+    }
+  }
+
+  if ( delay_dataset_exist_ )
+  {
+    try
+    {
+      delay_dset_ = edge_grp->openDataSet( "delay" );
+    }
+    catch ( const H5::Exception& e )
+    {
+      throw KernelException( "Could not open delay dataset in " + cur_fname_ + ": " + e.getDetailMsg() );
+    }
+  }
+
+  // TODO: If present, ensure correct size of syn_weight and delay dsets. This might not be straightforward if there
+  // are multiple edge id groups
```

Comment:
This is not an issue as long as we only support a single edge group, or? Then I would remove this comment, since it gives the impression that the current implementation is not safe. Are the sizes currently checked elsewhere?